### PR TITLE
Removing local cache lookup from broker

### PR DESCRIPTION
**Pull request by trwalke**
Fixes #
Fix for https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/3613

**Changes proposed in this request**
Msal will no longer look in local cache for tokens when broker is enabled

**Testing**
Unit test.
Manual test

**Performance impact**
Negligable

**Documentation**
- [ ] All relevant documentation is updated.

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs:191]: To avoid re-testing Android / iOS, it would be safer to do this only on non-mobile

**Comment by trwalke** [comment]
@bgavrilMS it looks like this test requires locally cached tokens to function. Is this test no longer possible since we can't return local tokens in acquireTokenSilent? Would this need to be an integration test with the WAM broker?
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/31a3ae3343a40b6312da7cf4a950161fd6dbaa41/tests/Microsoft.Identity.Test.Unit/PublicApiTests/PublicClientApplicationTests.cs#L859

File: src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs:191]: nit: better to place this complicated `if` condition in a private function, `ShouldPersistAccessToken()`

File: tests/Microsoft.Identity.Test.Unit/RequestsTests/SilentRequestTests.cs
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit/RequestsTests/SilentRequestTests.cs:99]: Can we have a unit test where we mock the broker so that it matches how iOS broker behaves? Have a look at [BrokerTests.cs ](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/main/tests/Microsoft.Identity.Test.Unit/BrokerTests/BrokerRequestTests.cs#L444)

File: tests/Microsoft.Identity.Test.Unit/RequestsTests/SilentRequestTests.cs
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit/RequestsTests/SilentRequestTests.cs:136]: Is it worth splitting this 4 data row test into 2 tests? I think they would be more readable.

**Comment by bgavrilMS** [comment]
> @bgavrilMS it looks like this test requires locally cached tokens to function. Is this test no longer possible since we can't return local tokens in acquireTokenSilent? Would this need to be an integration test with the WAM broker?
> 
> https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/31a3ae3343a40b6312da7cf4a950161fd6dbaa41/tests/Microsoft.Identity.Test.Unit/PublicApiTests/PublicClientApplicationTests.cs#L859

Yes, this is a good point. But I don't think we have infrastructure for such a test, i.e. do we have guest users in the lab? I recommend doing a manual test for now and adding a task to https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/3375

**Comment by bgavrilMS** [comment]
> > @bgavrilMS it looks like this test requires locally cached tokens to function. Is this test no longer possible since we can't return local tokens in acquireTokenSilent? Would this need to be an integration test with the WAM broker?
> > https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/31a3ae3343a40b6312da7cf4a950161fd6dbaa41/tests/Microsoft.Identity.Test.Unit/PublicApiTests/PublicClientApplicationTests.cs#L859
> 
> Yes, this is a good point. But I don't think we have infrastructure for such a test, i.e. do we have guest users in the lab? I recommend doing a manual test for now and adding a task to #3375

I did some manual testing myself and it seems ok. I pushed a small update to the dev app.

File: src/client/Microsoft.Identity.Client/Internal/Requests/Silent/SilentRequest.cs
Comment (trwalke) [reply src/client/Microsoft.Identity.Client/Internal/Requests/Silent/SilentRequest.cs:51]: I think so, Im trying to get my iOS setup so I can test/fine tune the code.

Comment (trwalke) [reply src/client/Microsoft.Identity.Client/Internal/Requests/Silent/SilentRequest.cs:51]: Actually, the code is fine as is, isBrokereConfigured will be false for iOS because CanBrokerSupportSilentAuth() will return false.

File: tests/Microsoft.Identity.Test.Unit/TestBase.cs
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit/TestBase.cs:106]: nit: this is ios broker mock, not wam

File: src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs:191]: iOS broker gives MSAL back the RT, because iOS does not allow two apps to talk to each other in the background. So iOS broker is the only broker for which we have the flag `CanBrokerSupportSilentAuth` set to false. 

IMO, I'd prefer to have this logic at runtime rather than at build time, i.e. if (BrokerSupportsSilentAuth) { ... } but I understand that we may not want to create new broker just for this. Perhaps this flag is better to be set in PlatformProxy and not in Broker ? Thoughts @trwalke ?

Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs:191]: I support runtime check i.e. CanBrokerSupportSilentAuth rather than compile time check, if things change in the future regarding which OS supports the silent broker.

File: tests/Microsoft.Identity.Test.Unit/BrokerTests/BrokerRequestTests.cs
Comment (SameerK-MSFT) [inline tests/Microsoft.Identity.Test.Unit/BrokerTests/BrokerRequestTests.cs:444]: This may be needed, if we do runtime check instead of compile time check.

**Review by SameerK-MSFT — CHANGES_REQUESTED** [review]

File: tests/Microsoft.Identity.Test.Unit/BrokerTests/BrokerRequestTests.cs
Comment (trwalke) [reply tests/Microsoft.Identity.Test.Unit/BrokerTests/BrokerRequestTests.cs:444]: It is still there. Just in a new location (TestBase.cs)

File: src/client/Microsoft.Identity.Client/Internal/Requests/Silent/SilentRequest.cs
Comment (trwalke) [reply src/client/Microsoft.Identity.Client/Internal/Requests/Silent/SilentRequest.cs:63]: The `isBrokerConfigured` value does not know if the broker is installed but the, the `BrokerSilentStrategy` will check and return null if it is not available. The code will handle this scenario by throwing an exception.

**Comment by bgavrilMS** [comment]
@SameerK-MSFT @trwalke - please aim to close this
@pmaytak - can you please review it as well?

@trwalke - from my point of view this looks good. I am just not sure if we don't break the iOS scenario. If you are able to manually test iOS, I'm signing off.

File: src/client/Microsoft.Identity.Client/Internal/Requests/Silent/SilentRequest.cs
Comment (neha-bhargava) [inline src/client/Microsoft.Identity.Client/Internal/Requests/Silent/SilentRequest.cs:53]: Do we need this check? I guess it is for silent call and we removed that call.

File: src/client/Microsoft.Identity.Client/Internal/Requests/Silent/SilentRequest.cs
Comment (pmaytak) [inline src/client/Microsoft.Identity.Client/Internal/Requests/Silent/SilentRequest.cs:84]:
```suggestion
                _logger.Verbose("Attempting to acquire token using local cache.");

```

File: src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs
Comment (pmaytak) [inline src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs:287]: nit:
```suggestion
#if iOS
            return msalAccessTokenCacheItem != null;
#else
            return msalAccessTokenCacheItem != null && tokenSource != TokenSource.Broker;
#endif
```

**Review by pmaytak — APPROVED** [review]

File: src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs
Comment (trwalke) [reply src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs]: this flag is already set in PlatformProxy so we are good

**Comment by trwalke** [comment]
> @SameerK-MSFT @trwalke - please aim to close this @pmaytak - can you please review it as well?
> 
> @trwalke - from my point of view this looks good. I am just not sure if we don't break the iOS scenario. If you are able to manually test iOS, I'm signing off.

PR has been updated. just waiting for PR feedback/approvals

**Review by bgavrilMS — APPROVED** [review]
Approved assuming manual testing of iOS scenario is ok.

File: tests/Microsoft.Identity.Test.Unit/PublicApiTests/PublicClientApplicationTests.cs
Comment (SameerK-MSFT) [inline tests/Microsoft.Identity.Test.Unit/PublicApiTests/PublicClientApplicationTests.cs:861]: If you are ignoring the test, why do you want to keep the code?

**Review by SameerK-MSFT — APPROVED** [review]

**Comment by trwalke** [comment]
I am trying to merge this PR in but I am getting an error that does not make sense to me. The error is: 

Test method Microsoft.Identity.Test.Integration.Broker.RuntimeBrokerTests.WamUsernamePasswordRequestAsync_WithPiiAsync threw exception:
Microsoft.Identity.Client.MsalUiRequiredException: WAM Error

Error Code: 0
Error Message: Account has previously been signed out of this application.
Internal Error Code: 540940121

I cant seem to find this error in our SDK so it looks like it is from WAM. However, it does not make sense. My changes are only affecting local MSAL cache so not sure how calling ATS on WAM can generate this error. There is also an identical test that does not use PII that works fine too.

Test: WamUsernamePasswordRequestAsync_WithPiiAsync - fails
Test: WamUsernamePasswordRequestAsync - passes

Any thoughts @bgavrilMS @SameerK-MSFT @gladjohn?

image

**Comment by gladjohn** [comment]
@SameerK-MSFT does this test `WamUsernamePasswordRequestAsync_WithPiiAsync` not run locally?

File: tests/Microsoft.Identity.Test.Unit/PublicApiTests/PublicClientApplicationTests.cs
Comment (bgavrilMS) [reply tests/Microsoft.Identity.Test.Unit/PublicApiTests/PublicClientApplicationTests.cs:861]: Agreed, delete it.

**Pull request merged**
